shared/src/main/kotlin/com/egm/stellio/shared/model/ApiExceptions.kt

```diff
@@ -2,24 +2,164 @@ package com.egm.stellio.shared.model
+import org.springframework.http.HttpStatus
+import org.springframework.http.MediaType
+import org.springframework.http.ProblemDetail
+import org.springframework.http.ResponseEntity
+import org.springframework.web.server.ServerWebExchange
+import reactor.core.publisher.Flux
+import reactor.core.publisher.Mono
+import java.net.URI
+
+const val DEFAULT_DETAIL = "If you have difficulty finding the issue" +
+    "you can check common issues here https://stellio.readthedocs.io/en/latest/TROUBLESHOOT.html" +
+    "If you think this is a bug of the broker" +
+    "you can create an issue on https://github.com/stellio-hub/stellio-context-broker"
 
 sealed class APIException(
-    override val message: String
-) : Exception(message)
-
-data class InvalidRequestException(override val message: String) : APIException(message)
-data class BadRequestDataException(override val message: String) : APIException(message)
-data class AlreadyExistsException(override val message: String) : APIException(message)
-data class OperationNotSupportedException(override val message: String) : APIException(message)
-data class ResourceNotFoundException(override val message: String) : APIException(message)
-data class InternalErrorException(override val message: String) : APIException(message)
-data class TooComplexQueryException(override val message: String) : APIException(message)
-data class TooManyResultsException(override val message: String) : APIException(message)
-data class AccessDeniedException(override val message: String) : APIException(message)
-data class NotImplementedException(override val message: String) : APIException(message)
-data class LdContextNotAvailableException(override val message: String) : APIException(message)
-data class NonexistentTenantException(override val message: String) : APIException(message)
-data class NotAcceptableException(override val message: String) : APIException(message)
+    val type: URI,
+    @JsonIgnore
+    val status: HttpStatus,
+    open val title: String,
+    @JsonProperty("detail")
+    override val message: String = DEFAULT_DETAIL
+) : Exception(message) {
+    private val logger = LoggerFactory.getLogger(javaClass)
+
+    fun toErrorResponse(): ResponseEntity<ProblemDetail> {
+        logger.info("Returning error ${this.type} (${this.message})")
+        return ResponseEntity.status(status)
+            .contentType(MediaType.APPLICATION_JSON)
+            .body(
+                ProblemDetail.forStatusAndDetail(status, this.message).also {
+                    it.title = this.title
+                    it.type = this.type
+                }
+            )
+    }
+
+    fun toServerWebExchange(exchange: ServerWebExchange): Mono<Void> {
+        logger.info("Returning server web exchange error ${this.type} (${this.message})")
+        exchange.response.statusCode = this.status
+        exchange.response.headers[CONTENT_TYPE] = MediaType.APPLICATION_JSON_VALUE
+        val body = serializeObject(this)
```

Comment:
here, you are also serializing the inherited `Exception` class, which is not the expected thing.

btw, we should also get rid of this inherited class (in FP, errors should not be exceptions). not sure of the impacts, to be checked.